library/std/src/sys/process/unix/unix/tests.rs

```diff
@@ -74,3 +83,111 @@ fn test_command_fork_no_unwind() {
             || signal == libc::SIGSEGV
     );
 }
+
+fn fd_test_stdin(use_exec: bool) {
+    let (pipe_reader, mut pipe_writer) = io::pipe().unwrap();
+
+    let fd_num = libc::STDIN_FILENO;
+
+    let mut cmd = Command::new("cat");
+    cmd.stdout(Stdio::piped()).fd(fd_num, pipe_reader);
+
+    if use_exec {
+        unsafe {
+            cmd.pre_exec(|| Ok(()));
+        }
+    }
+
+    let mut child = cmd.spawn().unwrap();
+    let mut stdout = child.stdout.take().unwrap();
+
+    assert_ne!(cmd.last_spawn_was_posix_spawn().unwrap_or(false), use_exec);
+
+    pipe_writer.write_all(b"Hello, world!").unwrap();
+    drop(pipe_writer);
+
+    child.wait().unwrap();
+    assert_eq!(io::read_to_string(&mut stdout).unwrap(), "Hello, world!");
+}
+
+fn fd_test_swap(use_exec: bool) {
+    let (pipe_reader1, mut pipe_writer1) = io::pipe().unwrap();
+    let (pipe_reader2, mut pipe_writer2) = io::pipe().unwrap();
+
+    let num1 = pipe_reader1.as_raw_fd();
+    let num2 = pipe_reader2.as_raw_fd();
+
+    let mut cmd = Command::new("cat");
+    cmd.arg(format!("/dev/fd/{num1}"))
+        .arg(format!("/dev/fd/{num2}"))
+        .stdout(Stdio::piped())
+        .fd(num2, pipe_reader1)
+        .fd(num1, pipe_reader2);
+
+    if use_exec {
+        unsafe {
+            cmd.pre_exec(|| Ok(()));
+        }
+    }
+
+    pipe_writer1.write_all(b"Hello from pipe 1!").unwrap();
+    drop(pipe_writer1);
+
+    pipe_writer2.write_all(b"Hello from pipe 2!").unwrap();
+    drop(pipe_writer2);
+
+    let mut child = cmd.spawn().unwrap();
+    let mut stdout = child.stdout.take().unwrap();
+
+    assert_ne!(cmd.last_spawn_was_posix_spawn().unwrap_or(false), use_exec);
+
+    child.wait().unwrap();
+    // the second pipe's output is clobbered; this is expected.
+    assert_eq!(io::read_to_string(&mut stdout).unwrap(), "Hello from pipe 1!");
+}
+
+// ensure that the fd is properly closed in the parent, but only after the child is spawned.
+fn fd_test_close_time(use_exec: bool) {
+    let (_pipe_reader, pipe_writer) = io::pipe().unwrap();
+
+    let fd = pipe_writer.as_raw_fd();
+    let fd_path = format!("/dev/fd/{fd}");
+
+    let mut cmd = Command::new("true");
+    cmd.fd(123, pipe_writer);
+
+    if use_exec {
+        unsafe {
+            cmd.pre_exec(|| Ok(()));
+        }
+    }
+
+    // Get the identifier of the fd (metadata follows symlinks)
+    let fd_id = md_file_id(&fs::metadata(&fd_path).expect("fd should be open"));
+
+    cmd.spawn().unwrap().wait().unwrap();
+
+    assert_ne!(cmd.last_spawn_was_posix_spawn().unwrap_or(false), use_exec);
+
+    // After the child is spawned, our fd should be closed
+    match fs::metadata(&fd_path) {
+        // Ok; fd exists but points to a different file
+        Ok(md) => assert_ne!(md_file_id(&md), fd_id),
+        // Ok; fd does not exist
+        Err(_) => (),
+    }
+}
+
+#[test]
+fn fd_tests_posix_spawn() {
+    fd_test_stdin(false);
+    fd_test_swap(false);
+    fd_test_close_time(false);
+}
```

Comment:
Copying from the cfg in `unix.rs` `posix_spawn`

```suggestion
#[test]
#[cfg_attr(not(any(
    target_os = "freebsd",
    target_os = "illumos",
    all(target_os = "linux", target_env = "gnu"),
    all(target_os = "linux", target_env = "musl"),
    target_os = "nto",
    target_vendor = "apple",
    target_os = "cygwin",
)), ignore = "posix_spawn unsupported")]
fn fd_tests_posix_spawn() {
    fd_test_stdin(false);
    fd_test_swap(false);
    fd_test_close_time(false);
}
```